server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -2155,6 +2289,73 @@ public Snapshot copySnapshot(CopySnapshotCmd cmd) throws StorageUnavailableExcep
         }
     }
 
+    private boolean canCopyOnPrimary(List<Long> poolIds, Snapshot snapshot) {
+        List<Long> poolsToBeRemoved = new ArrayList<>();
+        for (Long poolId : poolIds) {
+            PrimaryDataStore dataStore = (PrimaryDataStore) dataStoreMgr.getDataStore(poolId, DataStoreRole.Primary);
+            if (isObjectNull(dataStore == null, poolsToBeRemoved, poolId)) continue;
+
+            SnapshotInfo snapshotInfo = snapshotFactory.getSnapshot(snapshot.getId(), poolId, DataStoreRole.Primary);
+            if (isSnapshotExistsOnPool(snapshot, dataStore, snapshotInfo)) continue;
+
+            VolumeVO volume = _volsDao.findById(snapshot.getVolumeId());
+            if (isObjectNull(volume == null, poolsToBeRemoved, poolId)) continue;
+            doesStorageSupportCopySnapshot(poolsToBeRemoved, poolId, dataStore, volume);
+        }
+        poolIds.removeAll(poolsToBeRemoved);
+        if (CollectionUtils.isEmpty(poolIds)) {
+            return false;
+        }
+        return true;
+    }
+
+    private void doesStorageSupportCopySnapshot(List<Long> poolsToBeRemoved, Long poolId, PrimaryDataStore dataStore, VolumeVO volume) {
+        if (!dataStore.getDriver().getCapabilities()
+                .containsKey(DataStoreCapabilities.CAN_COPY_SNAPSHOT_BETWEEN_ZONES_AND_SAME_POOL_TYPE.toString())
+                && dataStore.getPoolType() != volume.getPoolType()) {
+            poolsToBeRemoved.add(poolId);
+            logger.debug(String.format("The %s  does not support copy to %s between zones", dataStore.getPoolType(), volume.getPoolType()));
+        }
+    }
+
+    private boolean isSnapshotExistsOnPool(Snapshot snapshot, PrimaryDataStore dataStore, SnapshotInfo snapshotInfo) {
+        if (snapshotInfo != null) {
+            logger.debug(String.format("Snapshot [%s] already exist on pool [%s]", snapshot.getUuid(), dataStore.getName()));
+            return true;
+        }
+        return false;
+    }
+
+    private static boolean isObjectNull(boolean object, List<Long> poolsToBeRemoved, Long poolId) {
```

Comment:
Can this method be renamed? By its name it would be assumed its just a null check but its doing more than that